src/hera/workflows/_runner/util.py

```diff
@@ -0,0 +1,261 @@
+    """Parse a value to the correct type.
+
+    Args:
+        value: The value to parse.
+        key: The name of the kwarg.
+        f: The function to parse the value for.
+
+    Returns:
+        The parsed value.
+
+    """
+    if _is_str_kwarg_of(key, f) or _is_artifact_loaded(key, f) or _is_output_kwarg(key, f):
+        return value
+    try:
+        if os.environ.get("hera__script_annotations", None) is None:
+            return json.loads(value)
+
+        type_ = _get_unannotated_type(key, f)
+        loaded_json_value = json.loads(value)
+
+        if not type_:
+            return loaded_json_value
+
+        _pydantic_mode = int(os.environ.get("hera__pydantic_mode", _PYDANTIC_VERSION))
+        if _pydantic_mode == 1:
+            return parse_obj_as(type_, loaded_json_value)
+        else:
+            return TypeAdapter(type_).validate_python(loaded_json_value)
+    except (json.JSONDecodeError, TypeError):
+        return value
+
+
+def _get_type(type_: type) -> type:
+    if get_origin(type_) is None:
+        return type_
+    origin_type = cast(type, get_origin(type_))
+    if origin_type is Annotated:
+        return get_args(type_)[0]
+    return origin_type
+
+
+def _get_unannotated_type(key: str, f: Callable) -> Optional[type]:
+    """Get the type of function param without the 'Annotated' outer type."""
+    type_ = inspect.signature(f).parameters[key].annotation
+    if type_ is inspect.Parameter.empty:
+        return None
+    if get_origin(type_) is None:
+        return type_
+
+    origin_type = cast(type, get_origin(type_))
+    if origin_type is Annotated:
+        return get_args(type_)[0]
+
+    # Type could be a dict/list with subscript type
+    return type_
+
+
+def _is_str_kwarg_of(key: str, f: Callable) -> bool:
+    """Check if param `key` of function `f` has a type annotation of a subclass of str."""
+    func_param_annotation = inspect.signature(f).parameters[key].annotation
+    if func_param_annotation is inspect.Parameter.empty:
+        return False
+
+    type_ = _get_type(func_param_annotation)
+    return issubclass(type_, str)
+
+
+def _is_artifact_loaded(key: str, f: Callable) -> bool:
+    """Check if param `key` of function `f` is actually an Artifact that has already been loaded."""
+    param = inspect.signature(f).parameters[key]
+    return (
+        get_origin(param.annotation) is Annotated
+        and isinstance(get_args(param.annotation)[1], Artifact)
+        and get_args(param.annotation)[1].loader == ArtifactLoader.json.value
+    )
+
+
+def _is_output_kwarg(key: str, f: Callable) -> bool:
+    """Check if param `key` of function `f` is an output Artifact/Parameter."""
+    param = inspect.signature(f).parameters[key]
+    return (
+        get_origin(param.annotation) is Annotated
+        and isinstance(get_args(param.annotation)[1], (Artifact, Parameter))
+        and get_args(param.annotation)[1].output
+    )
+
+
+def _runner(entrypoint: str, kwargs_list: List) -> Any:
```

Comment:
This seems like a very central runner function. I wonder if it can go into something more "significant" than a util file. I don't want to pick too much on "util", to be clear, but organizing into an abstract util does not tell contributors much atm :(

Reply:
Yes I was also not 100% behind this, but I didn't want to have `runner` in the path again, which would end up with `patch(..._runner.runner._runner)` in tests